### ISSUE TITLE
Submission form

### ISSUE THREAD

**Issue by SteveALee** [w3c/wai-interactive-lists#6]
Part of the workflows, see #4

(note the current [form for Evaluation Tools](https://www.w3.org/WAI/ER/tools/submission.php) appears to be broken. See [process notes](https://www.w3.org/WAI/ER/tools/README))

This repository includes a working example of  Form Submission using a set of basic form controls:

- text
- check box
- check box group - adds a "required" check before submission
- radio button
- Select - option list and multiple select list

## Pages

- [List view](https://wai-interactive-lists.netlify.app/test-list) of the submitted data files
- [Form](https://wai-interactive-lists.netlify.app/test-form) to add a new submission

## User Workflow

1. Go to the [Form](https://wai-interactive-lists.netlify.app/test-form) and fill it in, then submit.
1. You see the resulting JSON in the browser - or an error
1. Check [GitHub PRs](https://github.com/w3c/wai-interactive-lists/pulls) for a new one. Can take a while so you might want to look at the Actions to see how its progressing
1. Look around the PR - hopefully the contents provide all the info needed
1. Merge the PR to add the submission (or not, as required)
1. Wait for the site to rebuild and then look at the [list view](https://wai-interactive-lists.netlify.app/test-list)

## Coding notes.

TLDR, the form `name` / `value` attributes define the data. Append name with `[]` to always use array in the data, even single elements (though that is probably a bit "weird"). The values are purely data, not displayable strings. The example shows one way to map the values to display strings using a .yml file.

### Submission Data 

The submission data files are stored in [_data/submissions](https://github.com/w3c/wai-interactive-lists/tree/main/_data/submissions) with filename being the submission ref.

It's data only, not any presentation, so the code for the list and form need to map the value to displayed text. But that is not complicated given a few conventions.

#### Relationship to the Submission Form

The data is a JSON version of the HTML form submission format and the content depends on how the form is coded. Each data field name comes from related form control's `name` attribute. Normally, the data value is the control's `value` attribute . In the case of checkboxes the data field is either missing when unchecked, or the `value` attribute if checked, or `"on"` if not value is provided.
 
The values will be arrays if:

- the `name` attribute ends in `[]`
- or multiple options in a select are chosen
- or more than one check box hve the same name

In addition to the form fields, four metadata fields are provided:

 ` "form_name",  "form_version", "submission_date", "submission_ref"`

- versioning using `form_version` is recommended to make life easier - if the form changes then increase the version.
- if the `submission_ref` is not supplied in the form, one is added. This is to support a later features of submission editing.

## Code

The [form](https://github.com/w3c/wai-interactive-lists/blob/main/content/test-form.html) is "liquid + html" and the [list](https://github.com/w3c/wai-interactive-lists/blob/main/content/test-list.html) is "liquid + markdown" as these seem most appropriate.

Two Jekyll utility includes are provided:

- sort-data-folder - provides sorted access to any _data folder, specifically used to access the submission JSON files
- submission-form - for declaring a form for use with the submissions. The button is not included for flexibility

The file [_data/strings.yml](https://github.com/w3c/wai-interactive-lists/blob/main/_data/strings.yml) provides the display strings for mapping from the data values and also allows labels to be shared between the list and the form. This will also provide a route to translations in future.

For example this code displays a Select value

```liquid
{{strings.option_label}}: {{strings[ submission.option]}}
```

and this is the form coding for a single option select list:

```liquid
    <label>{{strings.option_label}}:
        <select name="option">
            <option value="option_none">{{strings.option_none}}</option>
            <option value="option_one">{{strings.option_one}}</option>
            <option value="option_two">{{strings.option_two}}</option>
        </select>
    </label>
```

For text controls

```
data -> "new_tool_title": "blah blah"
form -> <input type="text" name="new_tool_title"/> 
list -> <span id="new_tool_title">{% submission.new_tool_title %}</span>
```

### Filters

The existing filter code should work with a little adjustment. Use the array naming to get arrays in the data as required. The examples don't use id attributes for clarity. 

## Architecture

- A Netlify function deals with the form submissions
  - converts HTML form data to JSON
  - invokes the GitHub action
  - requires a GitHub PAC key in order to access
- A GitHub Action creates the PR
- A [Playwright](https://playwright.dev) test script is provided. This will be especially useful when forms have a lot of required fields. It also acted as an exploration of how useful Playwright tests might be for the WAI website.

## Observations

Something is sorting the JSON data alphabetically. I think it is GitHub but will try to stop it happening

The use of the unique id for the file name avoids issues of using a field and two submissions having the same text. It's less readable though.

## Next steps

We need the submission to redirect to either a preview or an acknowledgement page.

Should add a reviewers so GitHub notifications are fired when PR is created. Could be done by customising the Action per repo

**Comment by SteveALee** [comment]
- [ ] Ask sysrequest on technical and policy requirements for the endpoint to submit to.
- [ ]  get a temp solution working in Netlify previews (eg using a function)

**Comment by SteveALee** [comment]
@michael-n-cooper 

Looking at the legacy evaluation tools JSON data I see a number of arrays. I'm not sure why these are arrays in many cases.. other than grouping checkboxs. Perhaps some are just a result of the way the form coding was done? It seems to be mixing data and presentation - never a good idea.

Can you shed any lights so we can check the above proposal will work for the new data. My feeling is the flat format is fine and If you require to import old data we can run a simple conversions.

Here's an example:

```json
{ 
    "title": "a11y-checker",
    "creator": "Muhannad Abdelrazek",
    "location": "https:\/\/github.com\/Muhnad\/a11y-checker",
    "release": "2016-12-01",
    "version": "1.0.0",
    "description": "a11y-checker is a warn about HTML Markup code accessibility issue, you can know the accessibility issue by an open console and see the warning.",
    "a11yloc": "",
    "update": "2018-04-24",
    "language": [
        "English"
    ],
    "guideline": [
        "<strong><abbr title=\"Web Content Accessibility Guidelines\">WCAG<\/abbr> 2.0 \u2014 <abbr title=\"World Wide Web Consortium\">W3C<\/abbr> Web Content Accessibility Guidelines 2.0<\/strong>",
        "<abbr title=\"Web Content Accessibility Guidelines\">WCAG<\/abbr> 1.0 \u2014 <abbr title=\"World Wide Web Consortium\">W3C<\/abbr> Web Content Accessibility Guidelines 1.0"
    ],
    "assists": [
        "Generating reports of evaluation results",
        "Providing step-by-step evaluation guidance"
    ],
    "automated": [
        "Single web pages"
    ],
    "technology": [
        "<abbr title=\"Hypertext Markup Language\">HTML<\/abbr>"
    ],
    "type": [
        "Command line tool"
    ],
    "onlineservice": [],
    "desktop": [],
    "mobile": [],
    "authoringtools": [],
    "browsers": [],
    "license": [
        "Open Source"
    ]
}
```

**Comment by mhansma96** [comment]
Some fields should be arrays I think because they can have multiple (we also add some fields like operating systems which can be multiple). 
The mix of data with html is something I noticed as well, very dirty in my opinion and also a reason why the filters don't work at the moment - cannot properly match "html" with "<abbr title=\"Hypertext Markup Language\">HTML<\/abbr>" so we will work on cleaning that data
. 

Fields that can only have one value for a filter, should be text fields of course, but when looking at our new filters I think a lot of them will be arrays still.

**Comment by SteveALee** [comment]
> Some fields should be arrays I think because they can have multiple (we also add some fields like operating systems which can be multiple).
The mix of data with html is s

I agree in terms of good data modelling. But in the content of the code and data we have, I think a flattened set of separate values (checkboxes) also works. My thought is they are grouped using HTML semantics, eg field set in the form and ??? in the list.

If we use arrays then:

- I think arrays only apply to groups of checkboxes? is that correct
- We'd then need to map the flat form data to an array. Ican only  think that will mean adding Javascript that can recognise such a group and pass the array on to the Netlify function processing the submission. That will probably require generating the JSON in browser , or something more hacky like altering the forma data values to mark a group. After exploring some of the  options, I prefer just using the default form handling. Les code == less bugs, and especially less client side code is good.
- The form and list coding gets more complicated - OK but liquid is a bit rubbish with arrays

So the suggestion is that rather than

```json
{
"assists": [
        "reports",
        "step-by-step"
    ],
}
```

we use

```json
{
"assists_reports": "on",
"assists_step-by-step": "on",
}
```

**Comment by mhansma96** [comment]
I want to avoid using html semantics in the fields as this makes validating for filters a lot more difficult so I feel like we should avoid it, adding to this that it leads to the values only being able to be presented in the way that this html code alows it to be.

The array part is correct: only for groups of checkboxes and not for any other "normal" field. The reason it makes sense to makes them into arrays I think is that also in matching with filters this is the way it was set up in the template, I used Leticia's functions for this that already worked really well on mapping values of the checkboxes for one specific filter to the values in an array in the JSON for that specific filterfield. So maybe in the liquid it's an extra step as I noticed to show them, in the filtering it seems to make sense to use arrays, only when needed of course

**Comment by mhansma96** [comment]
Okay so to be clear in a situation like this:
![image](https://user-images.githubusercontent.com/84068842/153376539-d4e6e496-7b2e-43ea-a0ba-17c7aeb4561f.png)

A tool can work in multiple browsers so "correct" data management will tell us to save it like this:
`"browsers": ["Chrome", "Safari"]`

The filter-functions will also interpret this correctly and editting this or finding which browsers this tool work for within the JSON-file will also be a lot easier since it's grouped in this array. 
So that's okay right?

**Comment by SteveALee** [comment]
> I want to avoid using html semantics in the fields as this makes validating for filters a lot more difficult so I feel like we should avoid it, adding to this that it leads to the values only being able to be presented in the way that this html code alows it to be.

My feeling is 
- the data fields should be pure data
- the form and list pages decide how to present the data, but also attach the data to the elements (id or name)
- the filters use the data values which are set as attributes to the various DOM nodes that can be hidden

Does that agree with your thoughts?

I agree the filtering code is really elegant! We just need to convert the groups to arrays (including groups of one). I think we can easily add that to the filter code. Especially if we use a name prefix convention to tie everyhting together.

**Comment by mhansma96** [comment]
> the data fields should be pure data

This! That's exactly what I was thinking when I looked into the JSON and found these fields. Totally agree because it also blocks the filter-matching and forces display in that way (for example like a bold text etc.). Data fields should be data to not have dirty stuff in there

**Comment by SteveALee** [comment]
> Okay so to be clear in a situation like this: ![image](https://user-images.githubusercontent.com/84068842/153376539-d4e6e496-7b2e-43ea-a0ba-17c7aeb4561f.png)
> 
> A tool can work in multiple browsers so "correct" data management will tell us to save it like this: `"browsers": ["Chrome", "Safari"]`
> 
> The filter-functions will also interpret this correctly and editting this or finding which browsers this tool work for within the JSON-file will also be a lot easier since it's grouped in this array. So that's okay right?

So yes, okay in terms of data modelling and the filter code. But not a match for HTML forms which have no arrays! 

So we either have to use code to change the form submission or change the display and filtering code. I suggest naming convention to tie grouped items together.

**Comment by mhansma96** [comment]
> Oh, hold on!!!! My bad. Arrays of checkboxes DO exist in forms. I just remembered. not well documented but if several have the same name and different values, i'm pretty sure you get an array. D'oh!

Ah I see, we'll get there! Let me know if this is a route we can take then otherwise we'll revisit our options.
